lgc/patch/MeshTaskShader.cpp

```diff
@@ -541,4 +1054,1365 @@ Value *MeshTaskShader::convertToDivergent(Value *value) {
+                             });
+
+  // Primitive attribute export follows vertex attribute export
+  unsigned startLoc = inOutUsage.mesh.genericOutputMapLocCount;
+  for (auto &builtInExport : inOutUsage.mesh.builtInExportLocs) {
+    const unsigned exportLoc = builtInExport.second;
+    startLoc = std::max(startLoc, exportLoc + 1);
+  }
+
+  // Export primitive attributes (from generic outputs)
+  ldsStart = m_builder->getInt32(getMeshShaderLdsRegionStart(MeshLdsRegion::PrimitiveOutput));
+  auto primitiveStride = 4 * inOutUsage.perPrimitiveOutputMapLocCount;
+  auto ldsOffsetBase = m_builder->CreateMul(m_waveThreadInfo.threadIdInSubgroup, m_builder->getInt32(primitiveStride));
+  ldsOffsetBase = m_builder->CreateAdd(ldsStart, ldsOffsetBase);
+
+  for (unsigned loc = 0; loc < inOutUsage.mesh.perPrimitiveGenericOutputMapLocCount; ++loc) {
+    auto ldsOffset = m_builder->CreateAdd(ldsOffsetBase, m_builder->getInt32(4 * loc));
+    auto exportValue = readValueFromLds(FixedVectorType::get(m_builder->getFloatTy(), 4), ldsOffset);
+
+    std::array<Value *, 4> exportValues;
+    for (unsigned j = 0; j < 4; ++j)
+      exportValues[j] = m_builder->CreateExtractElement(exportValue, j);
+
+    m_builder->CreateIntrinsic(Intrinsic::amdgcn_exp, m_builder->getFloatTy(),
+                               {
+                                   m_builder->getInt32(EXP_TARGET_PARAM_0 + startLoc + loc), // tgt
+                                   m_builder->getInt32(0xF),                                 // en
+                                   // src0 ~ src3
+                                   exportValues[0], exportValues[1], exportValues[2], exportValues[3],
+                                   m_builder->getFalse(), // done
+                                   m_builder->getFalse(), // vm
+                               });
+    ++inOutUsage.primExpCount;
+  }
+
+  // Export primitive attributes (from built-ins as generic ones)
+  if (builtInUsage.primitiveId) {
+    if (inOutUsage.mesh.perPrimitiveBuiltInExportLocs.count(BuiltInPrimitiveId) > 0) {
+      assert(primitiveId);
+      auto undef = UndefValue::get(m_builder->getInt32Ty());
+      const unsigned exportLoc = inOutUsage.mesh.perPrimitiveBuiltInExportLocs[BuiltInPrimitiveId];
+
+      m_builder->CreateIntrinsic(Intrinsic::amdgcn_exp, m_builder->getInt32Ty(),
+                                 {
+                                     m_builder->getInt32(EXP_TARGET_PARAM_0 + startLoc + exportLoc), // tgt
+                                     m_builder->getInt32(0x1),                                       // en
+                                     // src0 ~ src3
+                                     primitiveId, undef, undef, undef,
+                                     m_builder->getFalse(), // done
+                                     m_builder->getFalse(), // vm
+                                 });
+      ++inOutUsage.primExpCount;
+    }
+  }
+
```

Comment:
Yes. I don't do this currently. I plan to do it when we support in/out packing for mesh shader. Current data structures are unable to support mesh shader in/out packing because we have per-primitive attributes.